File: YourBible/plugins/com-shortsands-audioplayer/src/ios/AudioPlayer/AudioReference.swift
//
//  AudioReference.swift
//  AudioPlayer
//
//  Created by Gary Griswold on 8/14/17.
//  Copyright © 2017 ShortSands. All rights reserved.
//

class AudioReference {
    
    let tocAudioBook: AudioTOCBook
    let chapter: String
    let fileType: String
    var audioChapter: AudioTOCChapter?
    
    init(book: AudioTOCBook, chapter: String, fileType: String) {
        self.tocAudioBook = book
        self.chapter = chapter
        self.fileType = fileType
        print("***** Init AudioReference ***** \(self.toString())")
    }
    
    convenience init(book: AudioTOCBook, chapterNum: Int, fileType: String) {
        let chapter = String(chapterNum)
        switch chapter.count {
        case 1:
            self.init(book: book, chapter: "00" + chapter, fileType: fileType)
        case 2:
            self.init(book: book, chapter: "0" + chapter, fileType: fileType)
        default:
            self.init(book: book, chapter: chapter, fileType: fileType)
        }
    }
   
    deinit {
        print("***** Deinit AudioReference ***** \(self.toString())")
    }

    var textVersion: String {
        get {
            return self.tocAudioBook.testament.bible.textVersion
        }
    }
 
    var silLang: String {
        get {
            return self.tocAudioBook.testament.bible.silLang
        }
    }

    var damId: String {
        get {
            return self.tocAudioBook.testament.damId
        }
    }
    
    var sequence: String {
        get {
            return self.tocAudioBook.bookOrder
        }
    }
    
    var sequenceNum: Int {
        get {
            return Int(self.tocAudioBook.bookOrder) ?? 1
        }
    }
    
    var bookId: String {
        get {
            return self.tocAudioBook.bookId
        }
    }
    
    var bookName: String {
        get {
            return self.tocAudioBook.bookName
        }
    }
    
    var chapterNum: Int {
        get {
            return Int(self.chapter) ?? 1
        }
    }
    
    var localName: String {
        get {
            return self.tocAudioBook.bookName + " " + String(Int(self.chapter) ?? 1)
        }
    }
    
    var dpbLanguageCode: String {
        get {
            return self.tocAudioBook.testament.dbpLanguageCode
        }
    }
    
    func nextChapter() -> AudioReference? {
        return self.tocAudioBook.testament.nextChapter(reference: self)
    }
    
    func priorChapter() -> AudioReference? {
        return self.tocAudioBook.testament.priorChapter(reference: self)
    }
    
    func getS3Bucket() -> String {
        switch (self.fileType) {
            case "mp3": return self.damId.lowercased() + ".shortsands.com"
            default: return "unknown bucket"
        }
    }
    
    func getS3Key() -> String {
        return self.sequence + "_" + self.bookId + "_" + self.chapter + "." + self.fileType
    }
    
    func getNodeId(verse: Int) -> String {
            return self.bookId + ":" + String(self.chapterNum) + ":" + String(verse)
    }
    
    func isEqual(reference: AudioReference) -> Bool {
        if (self.chapter != reference.chapter) { return false }
        if (self.bookId != reference.bookId) { return false }
        if (self.sequence != reference.sequence) { return false }
        if (self.damId != reference.damId) { return false }
        if (self.fileType != reference.fileType) { return false }
        return true
    }
    
    func toString() -> String {
        return self.damId + "_" + self.sequence + "_" + self.bookId + "_" + self.chapter + "." + self.fileType
    }
}

File: Plugins/AudioPlayer/src/android/app/src/main/java/com/shortsands/audioplayer/AudioReference.java
package com.shortsands.audioplayer;

/**
 * Created by garygriswold on 8/30/17.
 */
import android.util.Log;

class AudioReference {

    private static String TAG = "AudioReference";

    AudioTOCBook tocAudioBook;
    String chapter;
    String fileType;
    AudioTOCChapter audioChapter;

    AudioReference(AudioTOCBook book, String chapter, String fileType) {
        this.tocAudioBook = book;
        this.chapter = chapter;
        this.fileType = fileType;
        Log.d(TAG, "***** Init AudioReference ***** " + this.toString());
    }

    static AudioReference factory(AudioTOCBook book, int chapterNum, String fileType) {
        String chapter = String.valueOf(chapterNum);
        switch (chapter.length()) {
            case 1:
                return new AudioReference(book, "00" + chapter, fileType);
            case 2:
                return new AudioReference(book, "0" + chapter, fileType);
            default:
                return new AudioReference(book, chapter, fileType);
        }
    }

    //deinit {
    //    print("***** Deinit AudioReference ***** \(self.toString())")
    //}

    String textVersion() {
        return this.tocAudioBook.testament.bible.textVersion;
    }

    String silLang() {
        return this.tocAudioBook.testament.bible.silLang;
    }

    String damId() {
        return this.tocAudioBook.testament.damId;
    }

    String sequence() {
        return this.tocAudioBook.bookOrder;
    }

    Integer sequenceNum() { return this.tocAudioBook.sequence; }

    String bookId() {
        return this.tocAudioBook.bookId;
    }

    int chapterNum() {
        try {
            return Integer.parseInt(this.chapter);
        } catch(NumberFormatException ex) {
            return 1;
        }
    }

    String localName() {
        return this.tocAudioBook.bookName + " " + chapterNum();
    }

    String dbpLanguageCode() {
        return this.tocAudioBook.testament.dbpLanguageCode;
    }

    private String dbpVersionCode() { return this.tocAudioBook.testament.dbpVersionCode; }

    private String dbpBookName() { return this.tocAudioBook.dbpBookName; }

    AudioReference nextChapter() {
        return this.tocAudioBook.testament.nextChapter(this);
    }

    AudioReference priorChapter() {
        return this.tocAudioBook.testament.priorChapter(this);
    }

    String getS3Bucket() {
        if (this.fileType.equals("mp3")) {
            return "dbp-prod";
        } else {
            return "unknown bucket";
        }
    }

    String getS3Key() {
        String abbr = this.dbpLanguageCode() + this.dbpVersionCode();
        String chapNum = this.chapter;
        if (this.bookId() != "PSA") {
            chapNum = "_" + chapNum.substring(1);
        }
        String name = this.dbpBookName().replace(" ", "_");
        name = name + "____________".substring(0, (12 - name.length()));
        String key = String.format("audio/%s/%s/%s__%s_%s%s.%s", abbr, this.damId(), this.sequence(),
                chapNum, name, this.damId(), this.fileType);
        Log.d("TAG", "Key=" + key);
        return key;
    }

    String getNodeId(int verse)  {
        if (verse > 0) {
            return this.bookId() + ":" + String.valueOf(this.chapterNum()) + ":" + String.valueOf(verse);
        } else {
            return this.bookId() + ":" + String.valueOf(this.chapterNum());
        }
    }

    boolean isEqual(AudioReference reference) {
        if (!this.chapter.equals(reference.chapter)) { return false; }
        if (!this.bookId().equals(reference.bookId())) { return false; }
        if (!this.sequence().equals(reference.sequence())) { return false; }
        if (!this.damId().equals(reference.damId())) { return false; }
        if (!this.fileType.equals(reference.fileType)) { return false; }
        return true;
    }

    public String toString()  {
        return this.damId() + "_" + this.sequence() + "_" + this.bookId() + "_" + this.chapter + "." + this.fileType;
    }
}
